# Практика

## Определение вероятности. События

Требуется сгенерировать необходимые выборки и произвести по ним расчеты

### Задача 1
Брошено две монеты. Найти вероятность того, что монеты выпали разными сторонами

In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as sts
import random

In [8]:
n = 1000000
df = pd.DataFrame(columns = ["coin1", "coin2"])
coin1 = sts.bernoulli(.5) # Бернули для монеты 1
coin2 = sts.bernoulli(.5) # Бернули для монеты 2
df["coin1"] = coin1.rvs(n) 
df["coin2"] = coin2.rvs(n)
p = len(df[df["coin1"]!=df["coin2"]])/df.shape[0]
p #вероятность того, что монеты выпали разными сторонами 

0.499625

### Задача 2
Брошено три монеты. Описать множество всех элементарных событий. Найти вероятности следующих событий:


1.   A = {не выпало ни одного герба}
2.   B = {выпало четное число гербов}
3.   C = {на третьей монете выпал герб}


In [9]:
n = 100000
bernoulli_rv1 = sts.bernoulli(0.5)
bernoulli_rv2 = sts.bernoulli(0.5)
bernoulli_rv3 = sts.bernoulli(0.5)
b1 = bernoulli_rv1.rvs(n)
b2 = bernoulli_rv2.rvs(n)
b3 = bernoulli_rv3.rvs(n)

df = pd.DataFrame (zip(b1, b2, b3), columns = ["m1","m2","m3" ])
print (df)
print ('не выпало ни одного герба', df [(df.m1 != 0) & (df.m2 != 0) & (df.m3 != 0)].shape[0]/n)
print ('выпало четное число гербов', df [df.m1 + df.m2 + df.m3 == 1].shape[0]/n)
print ('на третьей монете выпал герб', df [(df.m3 == 0)].shape[0]/n)

       m1  m2  m3
0       0   1   1
1       1   1   1
2       0   0   1
3       1   0   1
4       1   1   0
...    ..  ..  ..
99995   0   1   0
99996   0   1   0
99997   0   0   0
99998   1   0   1
99999   1   1   1

[100000 rows x 3 columns]
не выпало ни одного герба 0.12632
выпало четное число гербов 0.37485
на третьей монете выпал герб 0.49861


Вероятность событий:
A = 0.12632
B = 0.37485
C = 0.49861

Предлагается также решить задачи: 

### Задача 3
Из двух претендентов E и L на ответственную должность три члена комиссии должны отобрать одного. Каждый член комиссии должен указать либо одного достойного, либо забраковать обоих. Претендент считается выбранным, если он был признана достойным хотя бы двумя членами комиссии.
Найти вероятность событий:

A = {рекомендован L}, B = {рекомендован E}


In [12]:
#какие варианты возможны:
a = [[0,1], [1,0], [0,0], [0,1], [1,0], [0,0]]
c= random.choices(a, k=3)

#есть три вероятности событий: рекомендован L,рекомендован E, нет рекомендации 
n = 3
df = pd.DataFrame(c, columns = ["E","L" ])
L = np.count_nonzero(df["L"])
E = np.count_nonzero(df["E"])
print (L)
print (E)

print (df)
print ('рекомендован L', df [(df.L >=2)].shape[1]/n)
print ('рекомендован E', df [(df.E >=2)].shape[1]/n)

1
1
   E  L
0  0  1
1  0  0
2  1  0
рекомендован L 0.6666666666666666
рекомендован E 0.6666666666666666


### Задача 4
Брошено две игральных кости. Описать множество элементарных событий. Найти вероятности событий:

A = {вышло две "шестерки"}

B = {сумма выпавших очков не меньше 11}

C = {не выпала ни одна "шестерка"}

In [13]:
n=100000
values1 = []
values2 = []

for x in range(n):
    values1.append (np.random.randint(1, 7))
    values2.append (np.random.randint(1, 7))
         
df = pd.DataFrame(zip(values1, values2 ),columns = ["cub1","cub2"] ) 

print (df)

print ('вышло две "шестерки"', df [((df.cub1 ==6)&(df.cub2 ==6))].shape[0]/n)
print ('сумма выпавших очков не меньше 11', df [((df.cub1 + df.cub2)>=11)].shape[0]/n)
print ('не выпала ни одна "шестерка"', df [(df.cub1 != 6)&(df.cub2 != 6) ].shape[0]/n)

       cub1  cub2
0         1     2
1         5     1
2         5     5
3         6     1
4         5     3
...     ...   ...
99995     5     4
99996     1     1
99997     4     2
99998     4     3
99999     6     2

[100000 rows x 2 columns]
вышло две "шестерки" 0.02787
сумма выпавших очков не меньше 11 0.08364
не выпала ни одна "шестерка" 0.69402


## Вероятность суммы событий

### Задача 5
Брошены две игральные кости. Найти вероятность события D = {выпала хотя бы одна шестёрка}


In [14]:
n=10000
values1 = []
values2 = []

for x in range(n):
    values1.append (np.random.randint(1, 7))
    values2.append (np.random.randint(1, 7))
         
df = pd.DataFrame(zip(values1, values2 ),columns = ["cub1","cub2"] ) 

print (df)

print ('выпала хотя бы одна шестёрка',\
       df [((df.cub1 ==6)&(df.cub2 !=6))|((df.cub1 !=6)&(df.cub2 ==6))|((df.cub1 ==6)&(df.cub2 ==6)) ].shape[0]/n)

      cub1  cub2
0        1     4
1        1     6
2        5     5
3        4     5
4        5     6
...    ...   ...
9995     4     3
9996     1     5
9997     6     6
9998     3     2
9999     3     1

[10000 rows x 2 columns]
выпала хотя бы одна шестёрка 0.3032


Предлагается также решить задачи: 

### Задача 6
В телефонном номере три последние цифры стерлись. Считая, что все возможные значения стершихся цифр равновероятны, найти вероятность событий:

A = {Стерлись различные цифры},

B = {Стерлись одинаковые цифры},

C = {Среди стершихся цифр хотя бы две совпадают},

D = {Среди стершихся цифр хотя бы две различны}


In [15]:
n=100000
num1 = []
num2 = []
num3 = []

for x in range(n):
    num1.append (np.random.randint(0, 10))
    num2.append (np.random.randint(0, 10))
    num3.append (np.random.randint(0, 10))    
df = pd.DataFrame(zip(num1, num2, num3 ), columns = ["num1","num2", "num3"] ) 

print (df)

print ('Стерлись различные цифры',\
       df [(df.num1 != df.num2)& (df.num1 != df.num3)& (df.num2 != df.num3)].shape[0]/n)
print ('Стерлись одинаковые цифры',\
       df [(df.num1 == df.num2)& (df.num1 == df.num3)& (df.num2 == df.num3)].shape[0]/n )
print ('Среди стершихся цифр хотя бы две совпадают',\
       df [(df.num1 == df.num2)|(df.num1 == df.num3)|(df.num2 == df.num3)].shape[0]/n )
print ('Среди стершихся цифр хотя бы две различны',\
       df [(df.num1 != df.num2)| (df.num1 != df.num3)|(df.num2 != df.num3)].shape[0]/n )

       num1  num2  num3
0         1     8     1
1         2     7     6
2         1     8     9
3         8     4     2
4         9     5     6
...     ...   ...   ...
99995     0     8     2
99996     5     7     9
99997     9     5     8
99998     8     4     4
99999     1     5     9

[100000 rows x 3 columns]
Стерлись различные цифры 0.7219
Стерлись одинаковые цифры 0.01001
Среди стершихся цифр хотя бы две совпадают 0.2781
Среди стершихся цифр хотя бы две различны 0.98999


## Случайные величины

### Задача 7
В лотерее имеется 10 билетов, из которых один выигрышный. Размер выигрыша 10 ден. ед.; стоимость билета 1 ден ед. Найти закон распределения случайной величины X, равной чистому выигрышу участника лотереи, который вытаскивает билет первым.

In [16]:
import random
lst = [0,0,0,0,0,0,0,0,0,1]
n = len(lst)
random.shuffle(lst, random.random)
df = pd.DataFrame(lst)
print (df)
print ('вероятность выигрыша 10 ден.ед: ', df [df[0] ==1 ].shape[0]/n)
print ('вероятность проигрыша ', df [df[0] ==0 ].shape[0]/n)

   0
0  0
1  0
2  0
3  1
4  0
5  0
6  0
7  0
8  0
9  0
вероятность выигрыша 10 ден.ед:  0.1
вероятность проигрыша  0.9


Предлагается также решить задачу: 

### Задача 8
Брошены две игральные кости. Найти закон распределения случайной величины $X$, равной сумме выпавших очков. Найти вероятности событий ${X <= 4}$, ${X > 4}$

In [17]:
n=10000
values = []
for x in range(n):
    values.append(np.random.randint(1, 7) + np.random.randint(1, 7))
df = pd.DataFrame(values) 

print(df)
print ('вероятности событий  𝑋<=4 ', df [df[0] <= 4 ].shape[0]/n)
print ('вероятности событий  𝑋>4 ', df [df[0] > 4 ].shape[0]/n)

       0
0     11
1      8
2      7
3      6
4      3
...   ..
9995   7
9996   8
9997   8
9998   3
9999   3

[10000 rows x 1 columns]
вероятности событий  𝑋<=4  0.1676
вероятности событий  𝑋>4  0.8324


## Условные вероятности

### Задача 9
Из колоды (36 карт) вынимают последовательно без возвращения две карты. Найти вероятность того, что первой картой была шестерка, а второй — семерка. Найти условную вероятность того же события при условии, что обе карты бубновой масти.


In [20]:
#Всего есть 9 карт бубновой масти, поэтому вероятность сначала 9/36, затем 8/35, нужно умножить их, чтобы расчитать 
#вероятность вытянуть их друг за другом

n1 = 36
k1= 9
P1 = k1 / n1

n2 = 35
k2 = 8
P2 = k2 / n2

p = P1 * P2
print(p)

0.05714285714285714


Предлагается также решить задачу: 

### Задача 10
Из урны, содержащей два белых и три черных шара, последовательно извлекают два шара; вынутые шары не возвращаются. Если первый шар оказался черным, то перед вторым извлечением в урну добавляют белый шар. В противном случае состав оставшихся шаров не меняют. Найти вероятности событий:

$A$ = {извлечены два белых шара},

$B$ = {извлечены два черных шара}

Указание. Пусть событие $C_1$ = {первый шар белый}, событие $C_2$ = {второй шар белый}. Тогда $A = C_1C_2$, $B = \overline C_1 \overline C_2$ 

In [21]:
#p1  = {извлечены два белых шара},
c1 = 2 / 5
c2 = 1 / 4
p1=c1*c2
print(p1)

0.1


In [22]:
#p2  = {извлечены два черных шара}
c1 = 3 /5
c2= 2 /5
p2 = c1*c2
print(p2)

0.24
